### ISSUE TITLE
Simplify and minimize module requirements

### ISSUE THREAD

**Issue by jywarren** [publiclab/image-sequencer#122]
Related to #242: first, in #122, we should minimize things we require in a module. Then we should make a range of abilities available to modules in #242.

https://github.com/publiclab/image-sequencer/blob/master/CONTRIBUTING.md#draw says:

> When you have done your calculations and produced an image output, you are required to set this.output to an object similar to what the input object was, call callback(), and set options.step.output equal to the output DataURL

I wonder if the basic module could be simplified with some helper functions:

```js
module.exports = function ModuleName(options,UI) {
  options = options || {};
  options.title = "Title of the Module";
  UI.onSetup(options.step);
  var output;

  function draw(input,callback) {
    UI.onDraw(options.step);

    var output = /*do something with the input*/ ;

    this.output = output;
    options.step.output = output.src;
    callback();
    UI.onComplete(options.step);
  }

  return {
    options: options,
    draw: draw,
    output: output,
    UI: UI
  }
}
```

First, I don't really like the double use of `var output` -- is there a reason we can't simply rely on the first definition of `output`? Do we need another one local to the `draw()` function, and if so could we name it something else?

Second, I wonder if these four lines could be in any way compressed:

```js
    this.output = output;
    options.step.output = output.src;
    callback();
    UI.onComplete(options.step);
```

Could we at least merge the first two? And is there any way we could just ask the module to run something like `done(output)` to just do everything, informing people that if they want to do something more unique, they can do these steps manually, but `done(output)` takes care of it in a standard default way? 

Just trying to think through this a bit @ccpandhare and see if we can make new module creation even a little simpler. Thanks!

****

### Checklist

* [ ] automate things like UI.onSetup(options.step);, UI.onDraw(options.step);, and UI.onComplete(options.step);.
* [ ] disambiguate/automate `this.output = output;` vs `options.step.output = output.src;`
* [ ] see if we can have a standard `output(src, format)` method which we simply pass everything to, which does these extra steps in a standard way
* [ ] automatically handle callback?
* [ ] update documentation in line with this example:

https://github.com/publiclab/image-sequencer/blob/5ac3ef008d177b3384aa26372ffc44194517718a/src/modules/Crop/Module.js#L34-L47

**Comment by jywarren** [comment]
Hmm, actually this one is much more readable! https://github.com/publiclab/image-sequencer/blob/master/src/modules/Crop/Module.js -- written later :-)

**Comment by ccpandhare** [comment]
Yes the Crop one is much more readable

**Comment by ccpandhare** [comment]
Yes the module creation is really complicated right now; What we can do is automate things like ` UI.onSetup(options.step);`, `UI.onDraw(options.step);`, and `UI.onComplete(options.step);`.

What do you think?

**Comment by jywarren** [comment]
That sounds pretty good. How do we break that project up so we can formulate a new module API?

**Comment by jywarren** [comment]
From #123 -- 

> Also - we should have a section in Contributing Modules that talks about Node/Browser compatibility with this in mind, so it's more clear to people, and we can also link to that section when people submit modules like this. We can have some possible solutions in that section too.

**Comment by ccpandhare** [comment]
I'm not sure about how we should go about the new API. But we do have to
think deeply about it to make it more developer friendly.

+1

**Comment by jywarren** [comment]
@ccpandhare @tech4GT I've consolidated some ideas for simplification above -- I think this could really dramatically simplify things! It could be like:

```js
module.exports = function ModuleName(options, UI) {
  options = options || {}; // automate this
  UI.onSetup(options.step); // automate this

  function draw(input,callback) {
    UI.onDraw(options.step); // automate this

    this.output = { 
      src: something(input.src); // do something to input,
      format: input.format
    }
    options.step.output = output.src; // automate this
    callback(); // automate this
    UI.onComplete(options.step); // automate this
  }

  return {
    options: options,
    draw: draw,
    output: output,
    UI: UI
  }
}
```

So then it'd be:

```js
module.exports = function ModuleName(options, UI) {
  function draw(input, callback) {
    output({
      src: doSomething(input.src); // do something to input,
      format: input.format
    })
  }
  return {
    options: options,
    draw: draw,
    output: output,
    UI: UI
  }
}
```

Or if we automatically pass the format (allowing it to be overwritten?):

```js
module.exports = function ModuleName(options, UI) {
  function draw(input, callback) {
    output(doSomething(input.src))
  }
  return {
    options: options,
    draw: draw,
    output: output,
    UI: UI
  }
}
```

What do you think?

**Comment by tech4GT** [comment]
@jywarren This looks amazing!!! Also I'll start making an issue for summer checklist, I'll link this there as well. Thanks 😁

**Comment by jywarren** [comment]
Just a note that I created #242 -- but to distinguish:

> First, in #122, we should minimize things we **require** in a module. Then we should make a range of abilities **available** to modules in #242.

Make sense?

Sorry to do all this fun stuff while you're in exams! I won't have as much time in the future, so I'm getting a bunch of things "set up" for when you start. Good luck, Varun!

**Comment by tech4GT** [comment]
@jywarren I was just reading #242, all of it makes complete sense, thanks😃
